File: C++/668.cpp
__________________________________________________________________________________________________
sample 8 ms submission
class Solution {
public:
    int GetLess(int m, int n, int P)
    {
        int beg = P/ n;
        int   t = n * (beg);
        for(int i = beg + 1; i <= m; ++ i)
        {
          t += (P / i); 
        }
        return t;
    }
    int findKthNumber(int m, int n, int k) {
        if(m > n) swap(m, n);
        int L = 1, R = m * n;
        while(L <= R)
        {
            int mid = (L + R) >> 1;
            int t = GetLess(m, n, mid);
            if(t >= k)
            {
                R = mid - 1;
            }
            else L = mid + 1;
        }
        return R+1;
    }
};
__________________________________________________________________________________________________
sample 8440 kb submission
class Solution {
public:
    bool is_valid(int x, int m, int n, int k) {
        int sum = 0;
        for (int i = 1; i <= m; i++) {
            sum += min(x/i, n);
        }
        return sum >= k;
    }
    int findKthNumber(int m, int n, int k) {
        int l = 1, r = m*n;
        while (l <= r) {
            int mid = l + r >> 1;
            if (is_valid(mid, m, n, k)) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return r + 1;
    }
};
__________________________________________________________________________________________________

File: Java/668.java
__________________________________________________________________________________________________
sample 6 ms submission
class Solution {
    public int findKthNumber(int m, int n, int k) {
        int l = 1;
        int r = m * n;
        while(l < r){
            int mid = l + ((r - l) >> 1);
            if(count(m, n, mid) >= k) r = mid;
            else l = mid + 1;
        }
        return l;
        /*
        PriorityQueue<int[]> pq = new PriorityQueue<int[]>(new Comparator<int[]>(){
            public int compare(int[] a, int[] b){
                return a[0] * a[1] - b[0] * b[1];
            }
        });
        for(int i = 1; i <= m; i++){
            pq.add(new int[]{i, 1});
        }
        for(int i = 0; i < Math.min(k - 1, m * n); i++){
            int[] curr = pq.poll();
            if(curr[1] < n)
                pq.add(new int[]{curr[0], curr[1] + 1});
        }
        return pq.peek()[0] * pq.peek()[1];
        */
    }
    private int count(int m, int n, int target){
        int counter = 0;
        /*
        for(int i = 1; i <= m > 0; i++){
            counter += Math.min(target / i, n);
        }
        */
        for(int i = 1, j = n; i <= m; i++){
            while(j >= 1 && i * j > target) j--;
            counter += j;
        }
        return counter;
    }
}
__________________________________________________________________________________________________
sample 31688 kb submission
class Solution {
    
    private int result;
    
    public int findKthNumber(int m, int n, int k) {
        if (m < n) {
            int t = m;
            m = n;
            n = t;
        }
        
        int a = 1;
        int b = m * n + 1;
        while (a <= b) {
            int sum = 0;
            int med = (a + b) / 2;
            result = 0;
            for (int i = 1; i <= n; i++) {
                int t = countLessThanOrEqual(med, i, m);
                sum += t;
                if (t == 0)
                    break;
            }
            
            if (sum == k || a == b)
                break;
            if (sum > k) {
                b = med;
            } else {
                a = med + 1;
            }
        }
        
        return result;
    }
    
    private int countLessThanOrEqual(int boundary, int a, int m) {
        int k = boundary / a;
        if (k > m)
            k = m;
        int t = a * k;
        if (t > result)
            result = t;
        return k;
    }
}
__________________________________________________________________________________________________
